lightning/src/offers/mod.rs

```diff
@@ -0,0 +1,541 @@
+// licenses.
+
+//! Implementation of Lightning Offers
+//! ([BOLT 12](https://github.com/lightning/bolts/blob/master/12-offer-encoding.md)).
+
+use bitcoin::bech32;
+use bitcoin::bech32::FromBase32;
+use bitcoin::blockdata::constants::genesis_block;
+use bitcoin::hash_types::BlockHash;
+use bitcoin::network::constants::Network;
+use bitcoin::secp256k1::{PublicKey, XOnlyPublicKey};
+use bitcoin::secp256k1::schnorr::Signature;
+use core::convert::TryFrom;
+use core::str::FromStr;
+use core::time::Duration;
+use ln::PaymentHash;
+use ln::features::OfferFeatures;
+use ln::msgs::DecodeError;
+use util::ser::{HighZeroBytesDroppedVarInt, Readable, WithLength, WithoutLength};
+
+use prelude::*;
+
+#[cfg(feature = "std")]
+use std::time::SystemTime;
+
+///
+#[derive(Clone, Debug)]
+pub struct Offer {
+	chains: Option<Vec<BlockHash>>,
+	amount: Option<Amount>,
+	description: String,
+	features: Option<OfferFeatures>,
+	absolute_expiry: Option<Duration>,
+	issuer: Option<String>,
+	destination: Destination,
+	quantity_min: Option<u64>,
+	quantity_max: Option<u64>,
+	recurrence: Option<Recurrence>,
+	send_invoice: Option<SendInvoice>,
+	signature: Option<Signature>,
+}
+
+impl Offer {
+	///
+	pub fn chain(&self) -> BlockHash {
+		self.chains
+			.as_ref()
+			.and_then(|chains| chains.first().copied())
+			.unwrap_or_else(|| genesis_block(Network::Bitcoin).block_hash())
+	}
+
+	///
+	pub fn amount(&self) -> Option<&Amount> {
+		self.amount.as_ref()
+	}
+
+	///
+	pub fn description(&self) -> &String {
+		&self.description
+	}
+
+	///
+	pub fn features(&self) -> Option<&OfferFeatures> {
+		self.features.as_ref()
+	}
+
+	///
+	pub fn absolute_expiry(&self) -> Option<Duration> {
+		self.absolute_expiry
+	}
+
+	///
+	#[cfg(feature = "std")]
+	pub fn is_expired(&self) -> bool {
+		match self.absolute_expiry() {
+			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
+				Ok(elapsed) => elapsed > seconds_from_epoch,
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	///
+	pub fn issuer(&self) -> Option<&String> {
+		self.issuer.as_ref()
+	}
+
+	///
+	pub fn destination(&self) -> &Destination {
+		&self.destination
+	}
+
+	///
+	pub fn quantity_min(&self) -> u64 {
+		self.quantity_min.unwrap_or(1)
+	}
+
+	///
+	pub fn quantity_max(&self) -> u64 {
+		self.quantity_max.unwrap_or_else(||
+			self.quantity_min.map_or(1, |_| u64::max_value()))
+	}
+
+	///
+	pub fn signature(&self) -> Option<&Signature> {
+		self.signature.as_ref()
+	}
+
+	fn to_tlv_stream(&self) -> OfferTlvStream {
+		let (currency, amount) = match self.amount.clone() {
+			None => (None, None),
+			Some(Amount::Bitcoin { amount_msats }) => (
+				None, Some(HighZeroBytesDroppedVarInt(amount_msats))
+			),
+			Some(Amount::Currency { iso4217_code, amount }) => (
+				Some(WithoutLength(iso4217_code)), Some(HighZeroBytesDroppedVarInt(amount))
+			),
+		};
+
+		let (paths, node_id) = match self.destination.clone() {
```

Comment:
@TheBlueMatt may be put off by Vec clones like these so we can do our best for devices that can't take too many memory allocations. In 1503 I found that disallowing extra vec allocations could sometimes change the code materially, so may be worth hashing out which vecs need to go early on.

Reply:
That was an advantage of (previously) having `Offer` wrap `OfferTlvStream` (no conversion needed), whereas the current approach trades that for having `Offer` always being in a correct state. I think I'm convinced that the non-wrapping version is better. But it would be nice to get rid of this drawback while still having the advantages mentioned in https://github.com/lightningdevkit/rust-lightning/pull/1597/files#r934961472.

Let's consider the use cases:
1. User scans QR code which gets parsed as an `Offer` via `OfferTlvStream` (no duplicate allocs)
2. User constructs an Offer programmatically to serialize as a string for a QR code (duplicate allocs)

The allocs could be removed if instead of using `Display` (which takes `&self`), we have a `From` conversion (which takes `self`) to `OfferTlvStream` or some other user-exposable type like `OfferString` or something. Though they may need to store the `Offer` as well (?), meaning they would need to copy it prior to serialization. Maybe serialization for QR code and in-memory storage would ultimately be different processes? We may need to put some more thought into how this may relate to `offer_id`s and to responding to `invoice_request`s. Let me re-read that part of the spec, but do let me know if you have any thoughts on this already.

Reply:
Pushed a couple commits at the end for signature verification and (related) `order_id` computation. Another point to consider is that -- while the `order_id` is computed as part of signature verification and can be stored in `Offer` when parsed -- for programmatic constructions, the `order_id` still needs to be computed by converting to the serialized TLV encoding either on-demand or at construction time. So we're going to get hit by allocations there it would seem regardless. Probably could reduce the number of allocations in merkle root computation, though.

Reply:
> Not sure I'm getting how this could relate to `offer_id`s, is there some way to clarify?

Mostly pointing out when responding to an `invoice_request`, we need to know if we actually know about an `Offer` with the corresponding `offer_id` from the request. So we'll need some way of determining this and responding with an appropriate `invoice`. I guess only tangentially related to allocs.
 
> For `invoice_request`s, I think it could make sense to have something like `Offer.get_invoice_request(amount, quantity, etc) -> Result<InvoiceRequest, Error>`. I was originally thinking this would live in `OnionMessenger` but it probably makes more sense in this module?

Yeah, that was going to be the next step here.

Reply:
> Pushed a couple commits at the end for signature verification and (related) `offer_id` computation. Another point to consider is that -- while the `offer_id` is computed as part of signature verification and can be stored in `Offer` when parsed -- for programmatic constructions, the `offer_id` still needs to be computed by converting to the serialized TLV encoding either on-demand or at construction time. So we're going to get hit by allocations there it would seem regardless. Probably could reduce the number of allocations in merkle root computation, though.

Would have to check it out more closely, mostly just wanted to note this as something to consider.

>Mostly pointing out when responding to an invoice_request, we need to know if we actually know about an Offer with the corresponding offer_id from the request. So we'll need some way of determining this and responding with an appropriate invoice. I guess only tangentially related to allocs.

I think we'll need to add something similar to the `inbound_payment` module but for inbound invoices. Then we can do `Offer::verify("payment_secret", &invoice, &inbound_invoice_secret)`. To be clear, the "payment_secret" comes in the `encrypted_data` we encoded for ourselves when we constructed the blinded route, so not 100% on where `verify` lives.